socs/agents/rtsp_camera/agent.py

```diff
@@ -0,0 +1,583 @@
+        curdaystr = f"{curtime.year}-{curtime.month:02d}-{curtime.day:02d}"
+
+        # The datetimes for start/stop today
+        def _dt_convert(timestr):
+            tstr = f"{curdaystr}T{timestr}"
+            try:
+                tm = datetime.strptime(tstr, "%Y-%m-%dT%H:%M:%S%z")
+            except ValueError:
+                tm = datetime.strptime(tstr, "%Y-%m-%dT%H:%M:%S")
+                msg = f"Motion time '{timestr}' is not "
+                msg += "timezone-aware.  Assuming UTC."
+                self.log.warning(msg)
+                tm = tm.replace(tzinfo=timezone.utc)
+            return tm
+        start = _dt_convert(self.motion_start)
+        stop = _dt_convert(self.motion_stop)
+
+        if stop <= start:
+            # We are starting today and stopping tomorrow
+            stop += timedelta(days=1)
+
+        if curtime > start and curtime < stop:
+            return True
+        else:
+            return False
+
+    @ocs_agent.param("test_mode", default=False, type=bool)
+    def acq(self, session, params=None):
+        """acq(test_mode=False)
+
+        **Process** - Capture frames from an RTSP camera.
+
+        Args:
+            test_mode (bool):  Run the Process loop only once. Meant only
+                for testing.  Default is False.
+
+        Notes:
+            Individual frames are written to a circular disk buffer.  Metadata
+            about the captured images is stored in the session data.  The format
+            of this is::
+
+                >>> response.session['data']
+                {
+                    'address': 'camera-c1.simonsobs.org',
+                    'timestamp': 1701983575.123456,
+                    'path': '/ocs/cameras_rtsp/c1/img_2023-12-29T02:44:47+00:00.jpg',
+                }
+
+        """
+        pm = Pacemaker(1 / self.seconds, quantize=False)
+        pmgrab = Pacemaker(self.record_fps, quantize=True)
+
+        frames_per_snapshot = int(self.seconds * self.record_fps)
+
+        session.set_status("running")
+        self.is_streaming = True
+
+        # Open camera stream
+        if self.fake:
+            cap = FakeCamera()
+        else:
+            cap = cv2.VideoCapture(self.connection)
+        if not cap.isOpened():
+            self.log.error(f"Cannot open RTSP stream at {self.connection}")
+            return False, "Could not open RTSP stream"
+
+        # Tracking state of whether we are currently recording motion detection
+        detecting = False
+        detect_start = None
+        record_frames = int(self.record_fps * self.record_duration)
+        motion_detector = MotionDetector()
+
+        snap_count = 0
+        while self.is_streaming:
+            # Use UTC
+            timestamp = time.time()
+            data = dict()
+
+            for iframe in range(frames_per_snapshot):
+                pmgrab.sleep()
+                # Grab an image
+                _ = cap.grab()
+            success, image = cap.retrieve()
```

Comment:
Does this `grab` have to be within the loop or can it just grab at the end? From my understanding, this is grabbing 200 images for example but only retrieving the last one.

Reply:
This is a consequence of the way these cameras work.  You have to continuously grab images and extract the ones you want.